cAVS IPC driver, power & clock unit tests

andyross opened the pull request:
This adds a register interface and driver for the host IPC hardware (essentially the bidirectional interrupt trigger with the host PC), along with a test for it based on some rework of the python scripting submitted earlier in #41803 (which I'm about to close).

Commit:
soc/intel_adsp: Add a cavs_ipc driver to manage host IPC

This is a slightly higher level Zephyr device that manages the host
IPC device for applications.  There's an interface to make synchronous
and asynchronous calls, to receive commands via (interrupt context)
callbacks and emit async "done" notifications after processing is
complete.  It should work for pretty much any application
architecture.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/boards: Add intel_adsp board integration/smoke test

As Zephyr begins to absorb drivers for these platforms that had
previously been managed by the SOF app, there's a need for a rapid
board-specific smoke test to use during development.

This starts with the smp_boot_delay test (itself a unit test for a
SOF-derived feature) and adds a host IPC case (that needs to match
code in cavstool.py on the other side of the PCI bus!).

It will grow more features over time as needed.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/boards/intel_adsp/smoke: Add board hardware features smoke test

These are descended from a private collection of test rigs I've been
semi-curating.  Getting cache, clock and memory space setup consistent
is mildly complicated on these devices and we've had bugs in all these
areas.

+ Check cache/uncache memory space setup, and make sure that the
  incoherent cache is operating as expected.

+ Make sure all cores agree with clocking setup.

+ Quickly benchmark a two-instruction loop to detect messups with
  instruction caching.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented on soc/xtensa/intel_adsp/common/cavs_ipc.c:
```diff
@@ -0,0 +1,164 @@
+	 * using it with IRQ_CONNECT here produces a wild preprocessor
+	 * error.  Why?
+	 */
+	irq_connect_dynamic(irq, 0, (void *)z_cavs_ipc_isr, dev, 0);
```
No, that's how I had it.  It didn't build.  Frankly it doesn't matter for these platforms since dynamic IRQs are architectural, but something to investigate.

teburd replied:
Is this with the xtensa compiler in the sdk or with the proprietary job? worked fine for me with the zephyr sdk compiler as I wrote it. I just built the board test with it.

andyross replied:
It was just the SDK.  I'll go back and check, it's entirely possible I messed something up.  I have minimal patience for devicetree burps, to be perfectly honest.

andyross replied:
Turned out that it didn't like the left typecast in the ISR parameter, because it wants to use it to glue up a symbol name.   Fixed.

andyross commented:
Sneak in one more patch with a productized smoke test for each core's cache & clock state.

andyross commented on include/drivers/cavs_ipc.h:
```diff
@@ -0,0 +1,160 @@
+ * complete and return notification to the other side (via the TDA
+ * register) is desired immediately.  Returning false means that no
+ * return "DONE" interrupt will occur until cavs_ipc_complete() is
+ * called on this device at some point in the future.
```
Hm.... I think I dropped a change here somewhere.  Somewhere I wrote a note about the fact that async DONE interrupts on cAVS 1.5 don't actually work (sorta, it's complicated), but I don't see it in the code as pushed.  Will fix.

Commit:
kernel: Refactor SMP cpu initialization a bit

Things had gotten a little tangled in there so let's do some cleanup.

Remove the distressingly-special-purpose z_reinit_idle_thread() hook
(which existed to support secondary core bringup when
SMP_BOOT_DELAY=y), and just fold that into a generic z_init_cpu(),
which we can call in obvious and symmetric ways from main
initialization, z_smp_init(), and z_smp_start_cpu() (the now-official
programmatic hook for starting cpus).

Remove the "#if CONFIG_MP_NUM_CPUS > 1" exclusions.  These weren't
saving any code size and were propagating themselves into platform
layers trying to avoid build failures.

There are some "special" APIs added for SOF which need to go away in
favor of the newer/generic z_smp_start_cpu().  Collect them in one
place and put them under a "#ifdef CONFIG_SOF" to prevent them from
being used in Zephyr apps.

Move some function declarations that didn't have homes into
<kernel/thread.h>.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nashif commented on soc/xtensa/intel_adsp/common/cavs_ipc.c:
```diff
@@ -0,0 +1,164 @@
+/* Copyright (c) 2022 Intel Corporation
```
why put the driver under the SOC and not under drivers/ipc/....

andyross replied:
It's a SOC-specific device for now.  I'll see about adding an extra API layer above it that can be more portably implemented

lgirdwood replied:
ack - this needs to SOC and protocol generic so it can be used by any SOC. At at mid level we need something like (feel free to modify)
```c
int ipc_send_message(uint32_t addr, char *doorbell, size_t door_size, char *mailbox, size_t mail_size, int blocking);
int ipc_recv_message(uint32_t addr, int (callback_func*) (doorbell, door_size, mailbox, mailsize));
```
The higher level application API would in cSOF would be similar to what it is today, but it would do the protocol mapping and call the above zephyr APIs.

andyross replied:
Inbound in maybe an hour.  Really it just productizes the "simple" bits of this hardware (so no weird return/DONE interrupt) as a generic subsystem API and implements it with the device as it stands already.  Finishing the typing part, need to write a test and it'll go onto the end of this PR.

nashif commented on soc/xtensa/intel_adsp/common/include/cavs-ipc-regs.h:
```diff
@@ -0,0 +1,93 @@
+/* Inter Processor Communication: a distressingly heavyweight method
+ * for directing interrupts at software running on other processors.
+ * Used for sending interrupts to and receiving them from another
+ * device.  cAVS uses it to talk to the host and the CSME, ACE also
```
what is ACE?

Commit:
soc/intel_adsp: Rework core start/restart API

Remove the soc_relaunch_cpu() API.  There's now nothing that this does
that isn't already done by the kernel's own z_smp_start_cpu() API.  We
don't need a SOC-level API for this anymore.

Rename soc_halt_cpu() to soc_adsp_halt_cpu() to clarify the domain,
simplify by removing the synchronization (there's nothing to
synchronize!).  Also move its declaration and docs into the soc.h
header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/boards/intel_adsp/smoke: Add cpu halt/restart case

Add code to exercise soc_adsp_halt_cpu() and validate the cores can be
restarted with the kernel z_smp_start_cpu() API.

Note that this interface is for 1.8+ devices only.  On cAVS 1.5, the
core power is actually controlled by the host.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented on include/drivers/cavs_ipc.h:
```diff
@@ -0,0 +1,160 @@
+#ifndef ZEPHYR_INCLUDE_CAVS_IPC_H
+#define ZEPHYR_INCLUDE_CAVS_IPC_H
+
+#include <cavs-ipc-regs.h>
```
It can indeed, fixed

andyross commented:
Seems like this may be in review for another few days, so sneak in three more patches containing work on the MP power management and some smoke test additions

andyross commented:
Push a more general API in the final patch of the series.  All this does is add a Zephyr device model subsystem wrapper around the driver that exposes the minimal/simplest/most-easily-reimplementable-on-other-platforms subset[1] of the IPC functionality as a generic API called "inotice".[2]

So instead of using the underlying cavs_ipc_*() API, you can use inotice_*() wrappers which indirect through the .api pointer of the device to find the specific callbacks.  And I was pleased to see that as long as one is careful to use "INOTICE_DEV" (which is a macro that expands to the static devicetree-instantiated device object), the wrappers get folded by the optimizer and the generated code is identical.

[1] Basically just sending messages consisting of two dwords of data (with platform-dependent significant bits) and receiving them on the other side in an interrupt callback.  I dropped the DONE interrupts and async command completion, though because the underlying cavs_ipc driver is still there you can use those features directly on the same device pointer and they'll work fine, just not portably.

[2] I'm not calling it IPC.  "Inter-process communication" in my world mean semaphores and queues and pipes and locks.  They aren't a hardware device.

nashif commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -0,0 +1,300 @@
+    (hda, sd, dsp, hda_ostream_id, cavs15) = map_regs()
+    log.info(f"Detected cAVS {'1.5' if cavs15 else '1.8+'} hardware")
+
+    load_firmware(sys.argv[1])
```
Need some error checking here and some help text describing what this tool does, if I run it standalone without arguments, I get:
```
Traceback (most recent call last):
  File "/home/nashif/cavstool.py", line 336, in <module>
    asyncio.get_event_loop().run_until_complete(main())
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/nashif/cavstool.py", line 319, in main
    load_firmware(sys.argv[1])
IndexError: list index out of range
```

and it needs root, so without sudo I get:

```
ups1:zephyr(main): python3 ~/cavstool.py
Traceback (most recent call last):
  File "/home/nashif/cavstool.py", line 336, in <module>
    asyncio.get_event_loop().run_until_complete(main())
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/nashif/cavstool.py", line 316, in main
    (hda, sd, dsp, hda_ostream_id, cavs15) = map_regs()
  File "/home/nashif/cavstool.py", line 38, in map_regs
    with open(f"{pcidir}/power/control", "w") as ctrl:
PermissionError: [Errno 13] Permission denied: '/sys/bus/pci/devices/0000:00:0e.0/power/control'

```

andyross replied:
This got pulled out and submitted separately in #42020 where that polish got added.

lgirdwood commented on include/drivers/cavs_ipc.h:
```diff
@@ -6,36 +6,12 @@
 
 #include <kernel.h>
 #include <device.h>
+#include <drivers/inotice.h>
```
Is there any need for a drivers/cavs_ipc.h as I would expect the inotice header to define the operations that each SOC IPC driver would use ?

andyross replied:
cavs_ipc is a more featureful interface, with support for async "DONE" notification on command completion.  The driver supports the inotice interface, which is a subsystem that can be implemented portably by any driver.  But also has its own API for the more complicated stuff.

Is that "needed"?  I don't know, but the hardware is there and it works and we have a driver and a test.  Seems like we should probably leave it in if someone wants to be creative with this device in the future.  You could use it for talking to the CSME if someone wants to evolve the protocol, etc...

Alternatively we could move the header back into the soc include directory so it doesn't confuse anyone.

andyross replied:
Oh, and to be clear there is no overhead: the functions that are common between the interfaces are identical, the calls from cavs_ipc that are unused get optimized out.  The ISR is identical regardless.

andyross commented:
Rebase; the cavstool script merged separately.

Commit:
tests/intel_adsp: Move IPC commands into shared header

It's useful for tests other than the IPC case to be able to
communicate with the script on the host, so generalize the interface
and move it to tests.h.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/intel_adsp: Force cpu_halt case to run on CPU 0

This test obviously only works correctly if it's run from core 0 (the
only CPU that doesn't shut down).  It was true essentially by
accident, but add some cpu_mask trickery to force things.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/intel_adsp: MP core power fixups for older cAVS platforms

On cAVS before 2.5, core power was controlled by the host.  Add a
command to the cavstool.py script to allow us to do that under test
command so we can exercise multiprocessor startup/shutdown outside of
SOF.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Add hardware race workaround to cavstool

On cAVS 1.8 (specifically) there seems to be a propagation delay on
the IPC registers.  Hitting the TDA register to signal DONE too soon
after clearing the interrupt via TDR can cause the interrupt to be
dropped.  Merely polling for it to read back correctly isn't
sufficient, we need an actual sleep here.

(The behavior that a message won't send while an existing message is
in progress is actually a hardware feature that is new with 1.8.  My
guess is it's a little glitchy in its first version.)

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/intel_adsp: Add clock calibration test

Now that we have easy access to code on the host, it's trivial to
check the clock against host timestamps with high precision.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
So how is this looking?  Things are starting to back up behind it, is this something folks think can be merged?  If not, can I get some direction as to what needs to be changed?  If we can't get to a resolution on an appropriate abstraction, can we at least merge the hardware driver layer on top of which we (really SOF: it's SOF that needs the abstraction) can build something later?

andyross commented:
It seems like we're in a sort of paralysis loop here, with @lgirdwood asking for an abstracted AP, and @nashif and @MaureenHelm uncomfortable with my (admittedly minimal) existing attempt.

But... there are some practical concerns here:

1. I need to get this merged so I can continue submitting more work

2. I need **this driver** merged, or something very like it, because I'm now using the host communication to the loader script to be able to write unit tests for multiprocessor startup and clock control.  And we want those!

So, how about this: I remove the cavs_ipc.h header and inotice.h headers both from the public include directories and put them in as SOC-level APIs only.  I can use them from there, and in the future when we decide on an appropriately portable SOF interface we can either evolve it or implement the portable API in terms of cavs_ipc.

Will that unblock the merge here?

andyross commented:
As promised, here's a version without any new public APIs, limiting the cavs_ipc driver to a soc role.  And I added the patches that are now using it to effect unit testing on the hardware, to demonstrate the value of getting this merged independently of a design for a portable SOF abstraction.

kv2019i reviewed (changes requested):
Excellent looking series, really good new coverage to the tests!

Unfortunately a test run with SOF caught an error. I did a quick bisect and basicly "kernel: Refactor SMP cpu initialization a bit" breaks SOF test cases on cAVS2.5 using the second DSP core. Launch CPU fails. I'll try to debug this a bit more today.

UPDATE: a potential SOF side change is needed after this PR. I'll update later on this. If yes, we can proceed with the merge and do matching change on SOF side.

kv2019i reviewed (dismissed):
Ok to go ahead with this. SOF has to make a matching patch when this is merged (for 3.1), but this is ok.

teburd reviewed (dismissed):
Re-reviewing as this has change quite a lot in the 19 or so days since I last looked. This is a nice changeset, and I'll be needing something (this exact sort of IPC with cavstool.py changes to test the HDA streams soon.

teburd commented:
I now have HDA working with host in/out streams and all, it depends on this branch currently as I'm using the cavs_ipc.h mechanism to ensure proper ordering of operations between the host and dsp. Any updates on this?

Commit:
drivers/ipm: Add IPM_CALLBACK_ASYNC mode

Some IPM devices support asynchronous command processing, where
acknowledgment of an IPM message can be delayed while handling
happens in a context other than the ISR that invoked the callback.
Expose this via a kconfig that can be selected by the driver, and add
a new "complete" call (a zero-overhead stub on non-supporting devices)
to signal the end of message handling.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/ipm: Add ipm_cavs_host: host/DSP communication on adsp_intel

Intel Audio DSPs have "IPC" interrupt delivery and shared memory
window hardware.  The Sound Open Firmware project has historically
used the combination of these to implement a bidirectional
message-passing interface.  As it happens, this protocol is an
excellent fit for Zephyr's somewhat geriatric but still useful IPM
interface.

This implements a SOF-protocol-compatible transport that will
hopefully prove a bit more futureproof for non-Intel SOF
architectures.  It is a software-only device, built on top of the
underlying SOC APIs for the SRAM windows (in cavs-shim) and IPC
(cavs_ipc).

Note that SOF actually has two protocol variants (ipc3 and ipc4): in
both, the command header (passed as the "id" parameter in IPM) is sent
via the hardware doorbell register.  But in ipc4, the second hardware
scratch register is used to transmit the first four bytes of the
command before involving the SRAM window (in ipc3, it's ignored).
Both modes are supported by this driver, set IPM_CAVS_HOST_REGWORD to
choose the "ipc4" variant.

Finally: note that the memory layout for the windows in question is
inherited from SOF, and for compatibility (with both SOF and with the
offsets used by the host!) these can't be changed without major
surgery.  They're defined in kconfig, but should be treated as
read-only until we get a chance to rework the way Zephyr does its SRAM
window management (and probably in concert with the host drivers).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/boards/intel_adsp: Add ipm_cavs_host test

Add a fairly simple test of the IPM-over-IPC driver.  This hits all
the code, but works by implementing the host side of the protocol
partially in the C test code.  The message is sent with an initial
payload, and then IPC commands from the firmware copy the data over
into the "inbox" region to simulate data being sent via the host.
Then we make sure it lands correctly as if the host driver had done it
directly.

This requries a new command in the cavstool script that will copy a
word from the "outbox" region to the "inbox" region (both are just
different SRAM windows, conceptually no different than the way the
script is already managing log output), but no significant surgery.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
> Why wont we start using more explicit/descriptive include paths

No argument at all from me about where to put the header.  This one is where it is simply because it's a SOC-specific driver (i.e. in its own directory that doesn't exist on Zephyr builds for other platforms) and because "cavs" is in the filename already so it seemed like it was reasonably disambiguated.  For cross platform driver subsystems (like IPM), we do in fact have a curated tree in include/drivers/<subsys>/...

Mostly though, these discussions generally turn into big bikesheds, so I keep my head down and try to structure changes to have minimal impact.  Realistically we're always moving stuff around anyway so any decision made will be ephemeral anyway. :)

But sure: if someone wants the driver somewhere else just say so and I'll move it.

andyross commented on soc/xtensa/intel_adsp/common/cavs_ipc.c:
```diff
@@ -0,0 +1,158 @@
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void z_cavs_ipc_isr(const void *devarg)
```
Yes, if we want to use the cavs_ipc driver on top of non-host IPC blocks that share interrupts with each other.  There, you need to figure out what device is involved via some other means and then call into the cavs_ipc "ISR" to do the per-IPC work.

andyross commented on soc/xtensa/intel_adsp/common/cavs_ipc.c:
```diff
@@ -0,0 +1,158 @@
+		return false;
+	}
+
+	k_sem_init(&devdata->sem, 0, 1);
```
Synchronization right here is actually provided by the spinlock.  The semaphore is for blocking  (exactly) one thread in send_message_sync().

andyross commented on soc/xtensa/intel_adsp/common/mp_cavs.c:
```diff
@@ -19,8 +19,6 @@
 
 #define CAVS15_ROM_IDC_DELAY 500
 
-static struct k_spinlock mplock;
```
It's provided by the OS now, arch_start_cpu() is unsynchronized.

andyross commented on tests/boards/intel_adsp/smoke/src/cpus.c:
```diff
@@ -0,0 +1,139 @@
+ */
+static ALWAYS_INLINE void delay_relax(void)
+{
+	for (volatile int j = 0; j < 1000; j++) {
```
Actually yes.  Or I've seen that happen historically (no idea if Zephyr/SOFs current toolchains and compile arguments produce that behavior).  And in any case, they're absolutely allowed to.  Volatile guarantees correct behavior.

marc-hb replied:
I remember having to find the optimizer every single time I tried to add code that "does nothing" (fortunately not that often).

On the other hand I'm surprised the test framework does not already have something like this, is this sort of test new to Zephyr?

andyross commented:
@lyakh I'm sorry, I came by to push another version and realized there were a bunch of stale comments to your review.  Please let me know if these don't resolve your concerns.  New version inbound in 3, 2, 1...

andyross reviewed (commented):
oops still didn't flush comments

andyross commented:
OK, new version up.  @lgirdwood what you will want to look at are the final three patches, which puts the SOF ipc3/ipc4 protocols under a Zephyr "IPM" device.  See API at https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/ipm.h (or generated doxygen in a link I'm having trouble finding at the moment, of course).

@teburd Your DMA PR on top of this is likely to collide a tiny bit (if nothing else with the command IDs for the cavstool script, because I added a new one), sorry.   Let me know if you have any trouble.

@aborisovich The broad idea here (that I **think** will work, but again this is based on merely reading the SOF ipc layer) is that you replace the platform_* layer under ipc3/4 with wrappers on this IPM device.  The first word of the command (the one passed in the TDR/IDR registers) goes in the IPM "id" parameter, the remaining ones are just passed as a byte array.  The new asynchronous mode added means that you can hold onto the received buffer (which is always the same inbox memory on cavs) and leave it in place as the result of mailbox_validate().  I do know ipc4 wants to leave this in place during processing (where ipc3 would copy it).  Hopefully there's no major surgery needed to SOF, just some wrappers.  And ideally there should be no changes needed to host drivers if I've done it correctly (which is an open question!).  Let me know if anything is confusing.

teburd commented:
No worries, I knew you were improving it, my patchset is relatively small and should be easy enough to rebase, will look closer at this soon

Commit:
CODEOWNERS: Add ipm_cavs_host

New driver needs owners (hijack the earlier cAVS wildcard for the
intra-DSP, it's now a legacy thing and essentially unrelated, but is
for the same platform).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Sorry, the push I just did was to clean up some missing licenses and add a CODEOWNERS line for the new driver.  I promise I'll yell if there are logic changes. :)

teburd commented on include/drivers/ipm.h:
```diff
@@ -221,6 +232,34 @@ static inline int z_impl_ipm_set_enabled(const struct device *ipmdev,
+ *
+ * @param ipmdev Driver instance pointer.
+ */
+__syscall void ipm_complete(const struct device *ipmdev);
```
Can I recommend allowing a pass through void* userdata as well here if that seems sensible, otherwise you are forcing static data to be used and you may wish to pass in a ipc object alongside the callback here without always creating custom callbacks.

andyross replied:
To be clear: that function pointer is a method, it's implemented by the driver, not the user of the API.

teburd replied:
Bah, ignore then, I commented too quickly on this

teburd reviewed (dismissed):
Looks good to me.
